mobile_verifier/src/heartbeats/last_location.rs

```diff
@@ -67,58 +119,197 @@ impl LocationCache {
-            )
-            .await;
+        let truncated = received_timestamp
+            .duration_trunc(Duration::hours(1))
+            .unwrap();
+        sqlx::query(
+            r#"
+            INSERT INTO wifi_heartbeats
+                (
+                    hotspot_key, location_validation_timestamp, latest_timestamp,
+                    truncated_timestamp, coverage_object,
+
+                    -- hardcoded values
+                    lat, lon, cell_type, distance_to_asserted, location_trust_score_multiplier
+                )
+            VALUES
+                (
+                    $1, $2, $3, $4, $5,
+
+                    -- harcoded values
+                    0.0, 0.0, 'novagenericwifiindoor', 0, 1000
+                )
+            "#,
+        )
+        .bind(hotspot)
+        .bind(validation_timestamp)
+        .bind(received_timestamp)
+        .bind(truncated)
+        .bind(Uuid::new_v4())
+        .execute(pool)
+        .await?;
+
         Ok(())
     }
 
-    /// Only used for testing.
-    pub async fn delete_last_location(&self, hotspot: &PublicKeyBinary) {
-        self.locations.remove(hotspot).await;
+    fn test_last_location(
+        latest_timestamp: DateTime<Utc>,
+        location_validation_timestamp: DateTime<Utc>,
+    ) -> LastLocation {
+        LastLocation {
+            location_validation_timestamp,
+            latest_timestamp,
+            lat: 0.0,
+            lon: 0.0,
+        }
+    }
+
+    #[sqlx::test]
+    async fn test_invalid_validation_timestamp(pool: PgPool) -> anyhow::Result<()> {
+        let now = Utc::now();
+
+        let hotspot_one = PublicKeyBinary::from(vec![1]);
+        let stale_timestamp = now - Duration::hours(24) - Duration::seconds(1);
+        insert_heartbeat(&pool, &hotspot_one, now, stale_timestamp).await?;
+
+        let hotspot_two = PublicKeyBinary::from(vec![2]);
+        let limit_timestamp = now - Duration::hours(24);
+        insert_heartbeat(&pool, &hotspot_two, now, limit_timestamp).await?;
+
+        let hotspot_three = PublicKeyBinary::from(vec![3]);
+        let good_timestamp = now - Duration::hours(12);
+        insert_heartbeat(&pool, &hotspot_three, now, good_timestamp).await?;
+
+        let cache = LocationCache::new(&pool);
+        assert_eq!(
+            None,
+            cache.get(&hotspot_one, now).await?,
+            "Invalid timestamp current"
+        );
+        assert_eq!(
+            Some(test_last_location(now, limit_timestamp)),
+            cache.get(&hotspot_two, now).await?,
+            "Limit timestamp current"
+        );
+        assert_eq!(
+            Some(test_last_location(now, good_timestamp)),
+            cache.get(&hotspot_three, now).await?,
+            "Good timestamp current"
+        );
+
+        // Moving an hour into the future should invalidate all timestamps
+        // regardless of what has already been cached.
+        let future = now + Duration::days(1);
```

Comment:
comment suggests moving 1 hour into future but it moves forward 1 day

Reply:
good eye, thank you